File: Java/src/main/java/QuestionBank/Solution001_100/Solution46.java
package QuestionBank.Solution001_100;

import java.util.ArrayList;
import java.util.List;


/**
 * @author windows
 */
public class Solution46 {
    public static void main(String[] args) {
        Solution46 solution46 = new Solution46();
        List<List<Integer>> permute = solution46.permute(new int[]{1, 2, 3});
        for (List list : permute) {
            System.out.println(list.toString());
        }
    }

    public List<List<Integer>> permute(int[] nums) {
        boolean[] used = new boolean[nums.length];
        List<List<Integer>> lists = new ArrayList<>();
        dfs(lists, new ArrayList<>(), nums, used);
        return lists;
    }

    private void dfs(List<List<Integer>> lists, List<Integer> list, int[] nums, boolean[] used) {
        if (list.size() == nums.length) {
            lists.add(new ArrayList<>(list));
            return;
        }
        for (int i = 0; i < nums.length; i++) {
            if (used[i] == false) {
                list.add(nums[i]);
                used[i] = true;
                dfs(lists, list, nums, used);
                used[i] = false;
                list.remove(list.size() - 1);
            }
        }
    }
}

File: C++/src/Solution1_100/Solution46.cpp
//
// Created by windows on 2023/3/2.
//

#include "Solution46.h"
#include <functional>
#include <iostream>
#include <vector>
#include <unordered_set>
#include <unordered_map>
#include <string>
using namespace std;
#include "vector"
using namespace std;

void dfs(vector<int>nums, vector<vector<int>>& res, vector<int>& path, vector<int>& used){
  if (path.size() == nums.size()){
    res.push_back(path);
    return;
  }
  for (int i = 0; i < nums.size(); i++){
    if (used[i] == 1) continue;
    path.push_back(nums[i]);
    used[i] = 1;
    dfs(nums, res, path, used);
    path.pop_back();
    used[i] = 0;
  }
}

vector<vector<int>> permute(vector<int>& nums) {
  vector<vector<int>> res;
  vector<int> path;
  vector<int> used(nums.size(), 0);
  dfs(nums, res, path, used);
  return res;
}
